src/pytorch_lightning/core/module.py

```diff
@@ -291,16 +290,24 @@ def _apply_batch_transfer_handler(
         self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0
     ) -> Any:
         device = device or self.device
-        datahook_selector = (
-            _DataHookSelector(self, None) if self._trainer is None else self.trainer._data_connector._datahook_selector
-        )
 
-        hook = datahook_selector.get_hook("on_before_batch_transfer")
-        batch = hook(batch, dataloader_idx)
-        hook = datahook_selector.get_hook("transfer_batch_to_device")
-        batch = hook(batch, device, dataloader_idx)
-        hook = datahook_selector.get_hook("on_after_batch_transfer")
-        batch = hook(batch, dataloader_idx)
+        def call_hook(hook_name, **kwargs):
+            if self._trainer:
+                datahook_selector = self._trainer._data_connector._datahook_selector
+                obj = datahook_selector.get_instance(hook_name)
+                trainer_method = (
+                    self._trainer._call_lightning_module_hook
+                    if isinstance(obj, self.__class__)
+                    else self._trainer._call_lightning_datamodule_hook
+                )
+                return trainer_method(hook_name, **kwargs)
+            else:
+                hook = getattr(self, hook_name)
+                return hook(**kwargs)
+
+        batch = call_hook("on_before_batch_transfer", batch=batch, dataloader_idx=dataloader_idx)
+        batch = call_hook("transfer_batch_to_device", batch=batch, device=device, dataloader_idx=dataloader_idx)
+        batch = call_hook("on_after_batch_transfer", batch=batch, dataloader_idx=dataloader_idx)
```

Comment:
I generally prefer to pass positional arguments positionally as it was done before. This will also avoid all those test updates.

Reply:
I think this is less error prone in case some updates are made to the API.
anyways, it doesn't affect the end user